src/kvstore.c

```diff
@@ -0,0 +1,589 @@
+        sum += kvs->state.dict_size_index[idx];
+        idx -= (idx & -idx);
+    }
+    return sum;
+}
+
+static void addDictIndexToCursor(kvstore *kvs, int didx, unsigned long long *cursor) {
+    if (kvs->num_dicts == 1)
+        return;
+    /* didx can be -1 when iteration is over and there are no more dicts to visit. */
+    if (didx < 0)
+        return;
+    *cursor = (*cursor << kvs->num_dicts_bits) | didx;
+}
+
+static int getAndClearDictIndexFromCursor(kvstore *kvs, unsigned long long *cursor) {
+    if (kvs->num_dicts == 1)
+        return 0;
+    int didx = (int) (*cursor & (kvs->num_dicts-1));
+    *cursor = *cursor >> kvs->num_dicts_bits;
+    return didx;
+}
+
+/* Updates binary index tree (also known as Fenwick tree), increasing key count for a given dict.
+ * You can read more about this data structure here https://en.wikipedia.org/wiki/Fenwick_tree
+ * Time complexity is O(log(kvs->num_dicts)). */
+static void cumulativeKeyCountAdd(kvstore *kvs, int didx, long delta) {
+    kvs->state.key_count += delta;
+
+    dict *d = kvstoreGetDict(kvs, didx);
+    size_t dsize = dictSize(d);
+    int non_empty_dicts_delta = dsize == 1? 1 : dsize == 0? -1 : 0;
+    kvs->state.non_empty_dicts += non_empty_dicts_delta;
+
+    /* BIT does not need to be calculated when there's only one dict. */
+    if (kvs->num_dicts == 1)
+        return;
+
+    /* Update the BIT */
+    int idx = didx + 1; /* Unlike dict indices, BIT is 1-based, so we need to add 1. */
+    while (idx <= kvs->num_dicts) {
+        if (delta < 0) {
+            assert(kvs->state.dict_size_index[idx] >= (unsigned long long)labs(delta));
+        }
+        kvs->state.dict_size_index[idx] += delta;
+        idx += (idx & -idx);
+    }
+}
+
+/**********************************/
+/*** dict callbacks ***************/
+/**********************************/
+
+/* Adds dictionary to the rehashing list, which allows us
+ * to quickly find rehash targets during incremental rehashing.
+ *
+ * If there are multiple dicts, updates the bucket count for the given dictionary
+ * in a DB, bucket count incremented with the new ht size during the rehashing phase.
+ * If there's one dict, bucket count can be retrieved directly from single dict bucket. */
+static void _dictRehashingStarted(dict *d) {
+    kvstoreDictMetadata *metadata = (kvstoreDictMetadata *)dictMetadata(d);
+    listAddNodeTail(metadata->kvs->state.rehashing, d);
+    metadata->rehashing_node = listLast(metadata->kvs->state.rehashing);
+
+    if (metadata->kvs->num_dicts == 1)
+        return;
+    unsigned long long from, to;
+    dictRehashingInfo(d, &from, &to);
+    metadata->kvs->state.bucket_count += to; /* Started rehashing (Add the new ht size) */
+}
+
+/* Remove dictionary from the rehashing list.
+ *
+ * Updates the bucket count for the given dictionary in a DB. It removes
+ * the old ht size of the dictionary from the total sum of buckets for a DB.  */
+static void _dictRehashingCompleted(dict *d) {
```

Comment:
```suggestion
static void dictRehashingCompleted(dict *d) {
```
I don't know why people started putting underscores in method names, we use static to indicate it's private most places.

Reply:
i did it because `dictMetadataSize` was already defined as a macro in dict.h

i'll add the `kvstore` prefix instead of an underscores